### PR TITLE
Sublinear memory cost policy

### PR DESCRIPTION
**Pull request by ir5**
Implemented sublinear memory cost policy from "Training Deep Nets with Sublinear Memory Cost (2016)". https://arxiv.org/abs/1604.06174

### CODE REVIEW

File: compiler/gradient_with_order.cc [hunk 1]
```diff
@@ -111,7 +125,9 @@ void AddGradientNodesForTrainingWithOrders(Graph* graph, const std::vector<Order
         last_forward_map[orig_node] = node;
         for (const auto& p : Zip(node->outputs(), orig_node->outputs())) {
             Value* value = std::get<0>(p);
-            CHECK(staged.emplace(std::get<1>(p), value).second);
+            // TODO(mkusumoto): Can I safely remove this assertion?
+            // CHECK(staged.emplace(std::get<1>(p), value).second);
```
Comment (shinh) [inline compiler/gradient_with_order.cc]: I think we'd like to get this CHECK back eventually. I'm fine with this TODO, but how about showing an error message like

```
            if (!staged.emplace(std::get<1>(p), value).second) {
                std::cerr << "Forward recompute without forgetting the output: " << orig_node->ToString() << std::endl;
            }
```

File: compiler/computation_order/policy_chen.cc [hunk 2]
```diff
@@ -0,0 +1,170 @@
+#include "compiler/computation_order/policy_chen.h"
```
Comment (shinh) [inline compiler/computation_order/policy_chen.cc]: Ah, how about mentioning a reference to the paper (https://arxiv.org/abs/1604.06174) somewhere in this file?

File: compiler/computation_order/core.h [hunk 3]
```diff
@@ -23,8 +24,38 @@ class Order {
...
+
+    friend std::ostream& operator<<(std::ostream& os, const Order& order);
 };
 
-std::vector<Order> GetComputationOrder(const Graph& graph, const std::string& policy);
+inline std::ostream& operator<<(std::ostream& os, const Order& order) {
+    os << "Order: ";
+    switch (order.kind) {
+        case Order::kComputeForward: {
+            os << "ComputeForward(node=" << order.node->outputs()[0]->name() << ")";
```
[the remark this answers is not in the document]
Comment (ir5) [reply compiler/computation_order/core.h]: Changed this part to use `ToString()`.

File: compiler/computation_order/core.h [hunk 4]
```diff
@@ -23,8 +24,38 @@ class Order {
...
     std::vector<int> indices;
+
+    Order(Kind kind_, Node* node_, Value* value_) : kind(kind_), node(node_), value(value_) {
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const Order& order);
 };
 
-std::vector<Order> GetComputationOrder(const Graph& graph, const std::string& policy);
+inline std::ostream& operator<<(std::ostream& os, const Order& order) {
```
[the remark this answers is not in the document]
Comment (ir5) [reply compiler/computation_order/core.h]: fixed.

File: compiler/computation_order/policy_chen.cc (hunk 2 above)
Comment (ir5) [reply compiler/computation_order/policy_chen.cc]: I referred the paper.

File: compiler/computation_order/policy_chen.cc [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+std::vector<Order> ChenPolicy(const Graph& graph) {
+    int64_t budget = g_chen_budget * 1000000LL;
+    if (g_chen_budget == 0) {
+        // default budget = sqrt of total memory
+        for (Node* node : graph.nodes()) {
+            for (Value* value : node->outputs()) {
+                budget += value->GetNBytes();
+            }
+        }
+        budget = budget / static_cast<int64_t>(std::sqrt(graph.nodes().size()));
```
[the remark this answers is not in the document]
Comment (ir5) [reply compiler/computation_order/policy_chen.cc]: fixed.

File: compiler/gradient_with_order.cc (hunk 1 above)
Comment (ir5) [reply compiler/gradient_with_order.cc]: I modified the code so that warning message is shown.

**Pull request merged**